mesa/experimental/devs/simulator.py

```diff
@@ -0,0 +1,265 @@
+
+    Attributes:
+        event_list (EventList): The list of events to execute
+        time (float | int): The current simulation time
+        time_unit (type) : The unit of the simulation time
+        model (Model): The model to simulate
+
+
+    """
+
+    # TODO: add replication support
+    # TODO: add experimentation support
+
+    def __init__(self, time_unit: type, start_time: int | float):
+        # should model run in a separate thread,
+        # and we can then interact with start, stop, run_until, and step?
+        self.event_list = EventList()
+        self.time = start_time
+        self.time_unit = time_unit
+        self.model = None
+
+    def check_time_unit(self, time: int | float) -> bool:
+        ...
+
+    def setup(self, model: "Model") -> None:
+        """Setup the model to simulate
+
+        Args:
+            model (Model): The model to simulate
+
+        Notes:
+            The basic assumption of the simulator is that a Model has a model.setup method that sets up the
+            model.
+
+        """
+
+        self.event_list.clear()
+        self.model = model
+        model.setup()
+
+    def reset(self):
+        """Reset the simulator by clearing the event list and removing the model to simulate"""
+        self.event_list.clear()
+        self.model = None
+
+    def run(self, time_delta: int | float):
+        """run the simulator for the specified time delta
+
+        Args:
+            time_delta (float| int): The time delta. The simulator is run from the current time to the current time
+                                     plus the time delta
+
+        """
+
+        end_time = self.time + time_delta
+        while self.time < end_time:
+            self.step()
+
+    def step(self):
+        event = self.event_list.pop_event()
+        self.time = event.time
+        event.execute()
+
+    def schedule_event_now(
+        self,
+        function: Callable,
+        priority: Priority = Priority.DEFAULT,
+        function_args: list[Any] | None = None,
+        function_kwargs: dict[str, Any] | None = None,
+    ) -> SimulationEvent:
+        """Schedule event for the current time instant
+
+        Args:
+            function (Callable): The callable to execute for this event
+            priority (Priority): the priority of the event, optional
+            function_args (List[Any]): list of arguments for function
+            function_kwargs (Dict[str, Any]):  dict of keyword arguments for function
+
+        Returns:
+            SimulationEvent: the simulation event that is scheduled
+
+        """
+
+        event = SimulationEvent(
+            self.time,
+            function,
+            priority=priority,
+            function_args=function_args,
+            function_kwargs=function_kwargs,
+        )
+        self._schedule_event(event)
+        return event
+
+    def schedule_event_absolute(
```

Comment:
Maybe `schedule_event_at`?

Reply:
I prefer relative and absolute because it is clearer to me.

Reply:
I think "now", "at" and "in" are more natural concepts of time. Beautiful Python reads like English. It would also be consistent with the current implementation.

https://github.com/projectmesa/mesa/blob/57d9cb565564cf8d4df9f68e5ca1ebb20dd99397/mesa/time.py#L472

@Corvince break the tie for us?

Reply:
Sorry, but I disagree with this fundamentally, "In" is very ambiguous because it can mean in a specific tick (so the same as "at"). Absolute and relative are just clear.

Reply:
Okay, let's go with the more explicit ones.